File: files/CSEC 05.05.py
import pandas as pd
from pandas.tseries.offsets import MonthEnd

input = pd.read_excel("files/Excel Challenge 5th May.xlsx", sheet_name="Sheet1", header=1, usecols="B:D")

input['Month'] = pd.to_datetime(input['Date']).dt.to_period('M').dt.to_timestamp() 
input['quarter'] = input['Month'].dt.quarter
input['year'] = input['Date'].dt.year

result = input.groupby(['year', 'Month', 'quarter']).agg({'Man Hour': 'sum', 'LTI Recorded': 'sum'}).reset_index()
result['valid'] = result['LTI Recorded'].gt(0).astype(int)
result['valid'] = result.groupby(['year', 'quarter'])['valid'].transform('all').astype(int)
result = result[["year", "quarter", "valid"]].drop_duplicates()
result = result.agg({'valid': 'sum'}).reset_index()

print(result) # 3

File: files/2024-05-12/CSEC 05.05.R
library(tidyverse)
library(readxl)

input = read_excel("files/Excel Challenge 5th May.xlsx", range = "B2:D41")

result = input %>%
  mutate(Month = floor_date(Date, "month"),
         quarter = quarter(Month), 
         year = year(Date)) %>%
  summarise(`Man Hour` = sum(`Man Hour`),
            `LTI Recorded` = sum(`LTI Recorded`),
            .by = c(year, Month, quarter)) %>%
  mutate(valid = ifelse(all(`LTI Recorded` > 0), 1, 0), .by = c(year, quarter)) %>%
  summarise(n = n_distinct(paste(year, quarter, sep = "-")), .by = valid) %>%
  filter(valid == 1) %>%
  pull(n)

print(result)
# [1] 3L
